### PR TITLE
feat: enable transaction simulations

### PR DESCRIPTION
**Pull request by OGPoyraz**
## **Description**

This PR introduces transaction simulation for the transactions.
Also include what's new modal changes.

In order to enable transaction simulations locally please use `MM_TRANSACTION_SIMULATIONS_ENABLED=true` in your env files.

For the detail of the patch please visit: https://github.com/MetaMask/core/pull/4348

## **Related issues**

Fixes: 
- https://github.com/MetaMask/mobile-planning/issues/1719
- https://github.com/MetaMask/mobile-planning/issues/1724

## **Manual testing steps**

See recording

## **Screenshots/Recordings**

https://github.com/MetaMask/metamask-mobile/assets/7644512/fbc5e60e-7c3d-4649-94b8-635807b2ece9

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/core/Engine.ts [hunk 1]
```diff
@@ -1068,6 +1068,7 @@ class Engine {
           });
         },
       },
+      isSimulationEnabled: () => true,
```
Comment (OGPoyraz) [inline app/core/Engine.ts:1071]: This is temporary for testing purpose.

File: app/components/UI/Name/Name.tsx [hunk 2]
```diff
@@ -1,21 +1,44 @@
...
+
+// The number of characters to slice from the beginning of an address for truncated format:
+// `${TRUNCATED_ADDRESS_START_CHARS}...${TRUNCATED_ADDRESS_END_CHARS}`
+const TRUNCATED_ADDRESS_START_CHARS = 7;
+
+// The number of characters to slice from the end of an address for truncated format:
+// `${TRUNCATED_ADDRESS_START_CHARS}...${TRUNCATED_ADDRESS_END_CHARS}`
+const TRUNCATED_ADDRESS_END_CHARS = 5;
+
+export function shortenAddress(address = '') {
```
Comment (matthewwalsh0) [inline app/components/UI/Name/Name.tsx:33]: As in the extension, should we put this and the constants in a util somewhere so it's not coupled to `Name` and can have its own unit tests?

File: app/components/Views/confirmations/SendFlow/Confirm/index.js [hunk 3]
```diff
@@ -1325,6 +1334,15 @@ class Confirm extends PureComponent {
               </View>
             </View>
           )}
+          {transactionState?.id && (
+            <View style={styles.simulationWrapper}>
+              <SimulationDetails
+                simulationData={transactionSimulationData}
+                enableMetrics={false}
```
Comment (matthewwalsh0) [inline app/components/Views/confirmations/SendFlow/Confirm/index.js:1341]: Do we not want metrics for transfer transactions?

File: app/selectors/confirmTransaction.ts [hunk 4]
```diff
@@ -8,7 +8,8 @@ import { mergeGasFeeEstimates } from '@metamask/transaction-controller';
 import { createSelector } from 'reselect';
 import { createDeepEqualSelector } from './util';
 
-const selectCurrentTransactionId = (state: RootState) => state.transaction?.id;
+export const selectCurrentTransactionId = (state: RootState) =>
```
Comment (matthewwalsh0) [inline app/selectors/confirmTransaction.ts:11]: Do we need to export this?

Could we instead just use the below `selectCurrentTransactionMetadata` and not make a simulation data specific selector, as that would mean we'd need selectors for every future metadata property also?

**Comment by sleepytanya** [comment]
Physical Samsung S24+, Android 14. [Bitrise build](https://app.bitrise.io/build/12a21b74-2cb0-449a-93e9-210eb844b599?tab=artifacts).

Simulations conducted on the Sepolia network perform as anticipated. For some reason the balance on Sepolia is always shown as 0 but that doesn't prevent the user to send tokens. `$NaN` is displayed within the simulation section next to the currency name: 


https://github.com/MetaMask/metamask-mobile/assets/104780023/00d6aa59-be68-4210-8e53-aa310fa3828e




When operating on the mainnet, the application experiences a significant slowdown, with transactions taking minutes to proceed. Confirmation screen for the swap displays a message 'You need *** more *token_name* to proceed...' while account has enough assets to pay for the transaction. Sometimes swap simulation work but it takes minutes for them to appear:

https://github.com/MetaMask/metamask-mobile/assets/104780023/95c5d8bc-04ca-4c98-90a4-fc9ed3b51138

 Simple Send on mainnet is also extremely slow but simulations work:

https://github.com/MetaMask/metamask-mobile/assets/104780023/84b7d521-f7d8-494f-a2da-6ec54d910850



I don't see simulations working for the test dapp - is it expected?

**Comment by OGPoyraz** [comment]
@sleepytanya concerns about the fiat values:
- 0 value on balance: I think you need to enable Advanced => Show conversion on test networks
- NaN value is the same reason as above, you need to enable the conversion on test networks. Since this `NaN US$` looks ugly I've already created a task to handle it separately (https://github.com/orgs/MetaMask/projects/78/views/1?pane=issue&itemId=65817155) but we didn't had a chance to estimate it yet.

I think there is overall a performance issue on your device somehow, I've never experienced such degradation on simulator/phyiscal device (could be seen from the video I recorded). But I will build and check it one more time.

File: app/core/Engine.ts (hunk 1 above)
Comment (OGPoyraz) [reply app/core/Engine.ts:1071]: > Could we use an ENV for this or will this be the final ticket?

I am thinking to add ENV for this, preference and the what's new PR. But to make it easy to test it for QA, I didn't add this up yet.

> Did we want to also add E2E tests before enabling?

This has separate ticket that I am currently working on https://github.com/MetaMask/mobile-planning/issues/1725

> We can also reference the preference once that is merged?

Exactly, I will first merge the pref go with ENV then update this.

File: app/components/Views/confirmations/SendFlow/Confirm/index.js (hunk 3 above)
Comment (OGPoyraz) [reply app/components/Views/confirmations/SendFlow/Confirm/index.js:1341]: Good catch, a forgotten one.

File: app/selectors/confirmTransaction.ts (hunk 4 above)
Comment (OGPoyraz) [reply app/selectors/confirmTransaction.ts]: Done https://github.com/MetaMask/metamask-mobile/pull/9828/commits/533fb74b4eb79c9e5d02d811e74e1596604a3cf2

File: app/core/Engine.ts (hunk 1 above)
Comment (OGPoyraz) [reply app/core/Engine.ts]: Update: toggle is merged, this is now referencing preferences controller state, but still hidden behind env var.

**Review by matthewwalsh0 — DISMISSED** [review]

**Review by vinistevam — DISMISSED** [review]

**Review by Cal-L — CHANGES_REQUESTED** [review]
Add `MM_TRANSACTION_SIMULATIONS_ENABLED` to `bitrisel.yml`. You can reference how `MM_NOTIFICATIONS_UI_ENABLED` is set

**Review by Cal-L — APPROVED** [review]
LGTM

**Pull request merged**